apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java

```diff
@@ -100,49 +122,96 @@ public static void onEnterServletService(@Advice.Argument(0) ServletRequest serv
+                transaction.setFrameworkName(FRAMEWORK_NAME);
+
+                servletTransactionHelper.fillRequestContext(transaction, request.getProtocol(), request.getMethod(), request.isSecure(),
+                    request.getScheme(), request.getServerName(), request.getServerPort(), request.getRequestURI(), request.getQueryString(),
+                    request.getRemoteAddr(), request.getHeader("Content-Type"));
+            } else if (transaction == null) {
+                final AbstractSpan<?> parent = tracer.getActive();
+                if (parent != null) {
+                    final HttpServletRequest request = (HttpServletRequest) servletRequest;
+                    DispatcherType dispatcherType = request.getDispatcherType();
+                    boolean isSpannableDispatcherType = false;
+                    span = parent.createSpan()
+                        .withType(SPAN_TYPE)
+                        .withSubtype(SPAN_SUBTYPE);
+
+                    if (dispatcherType == DispatcherType.FORWARD) {
+                        span.appendToName(FORWARD)
+                            .appendToName(SPACE)
+                            .appendToName(request.getServletPath());
+                        if (request.getPathInfo() != null) {
+                            span.appendToName(request.getPathInfo());
+                        }
+                        span.withAction(FORWARD_SPAN_ACTION);
+                        isSpannableDispatcherType = true;
+                    } else if (dispatcherType == DispatcherType.INCLUDE) {
+                        Object pathInfo = request.getAttribute(RequestDispatcher.INCLUDE_PATH_INFO);
+                        Object includeServletPath = request.getAttribute(RequestDispatcher.INCLUDE_SERVLET_PATH);
+                        span.appendToName(INCLUDE)
+                            .appendToName(SPACE);
+                        if (includeServletPath != null) {
+                            span.appendToName((String) includeServletPath);
+                        }
+                        if (pathInfo != null) {
+                            span.appendToName((String) pathInfo);
+                        }
+                        span.withAction(FORWARD_SPAN_ACTION);
+                        isSpannableDispatcherType = true;
+                    } else if (dispatcherType == DispatcherType.ERROR) {
+                        Object servletPath = request.getServletPath();
+                        span.appendToName(ERROR)
+                            .appendToName(SPACE);
+                        if (servletPath != null) {
+                            span.appendToName((String) servletPath);
+                        }
+                        span.withAction(ERROR_SPAN_ACTION);
+                        isSpannableDispatcherType = true;
+                    }
+                    if (isSpannableDispatcherType && !parent.getNameAsString().equals(span.getNameAsString())) {
```

Comment:
`getNameAsString` is allocating a String objects (so two in this comparison).
There are better ways to avoid creation of same nested spans. For example, before creating a span, see if the parent span has the same type, subtype and action. 
What cases are you trying to prevent with this comparison?